File: src/main/java/ru/hollowhorizon/hc/client/screens/util/Alignment.java
package ru.hollowhorizon.hc.client.screens.util;

public enum Alignment implements IPlacement {
    BOTTOM_CENTER(0.5F, 1.0F),
    BOTTOM_RIGHT(1F, 1.0F),
    BOTTOM_LEFT(0F, 1.0F),

    CENTER(0.5F, 0.5F),
    RIGHT_CENTER(1F, 0.5F),
    LEFT_CENTER(0F, 0.5F),

    TOP_CENTER(0.5F, 0F),
    TOP_RIGHT(1F, 0F),
    TOP_LEFT(0F, 0F);


    private final float factorX;
    private final float factorY;

    Alignment(float factorX, float factorY) {
        this.factorX = factorX;
        this.factorY = factorY;
    }

    public static IPlacement custom(float factorX, float factorY) {
        return new IPlacement() {

            @Override
            public float factorX() {
                return factorX;
            }

            @Override
            public float factorY() {
                return factorY;
            }
        };
    }

    @Override
    public float factorX() {
        return factorX;
    }

    @Override
    public float factorY() {
        return factorY;
    }
}

File: src/main/java/ru/hollowhorizon/hc/common/ui/Alignment.kt
/*
 * MIT License
 *
 * Copyright (c) 2024 HollowHorizon
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package ru.hollowhorizon.hc.common.ui

enum class Alignment(override val factorX: Float, override val factorY: Float): IPlacement {
    BOTTOM_CENTER(0.5F, 1.0F),
    BOTTOM_RIGHT(1F, 1.0F),
    BOTTOM_LEFT(0F, 1.0F),

    CENTER(0.5F, 0.5F),
    RIGHT_CENTER(1F, 0.5F),
    LEFT_CENTER(0F, 0.5F),

    TOP_CENTER(0.5F, 0F),
    TOP_RIGHT(1F, 0F),
    TOP_LEFT(0F, 0F);

    fun factorX() = factorX
    fun factorY() = factorY
}


enum class Anchor(val factor: Float) {
    START(0f),
    CENTER(0.5f),
    END(1f)
}


interface IPlacement {
    val factorX: Float
    val factorY: Float
}

